keps/sig-storage/3294-provision-volumes-from-cross-namespace-snapshots/README.md

```diff
@@ -0,0 +1,1024 @@
+
+### API
+
+A new `VolumeSnapshotLink` CRD is introduced in `snapshot.storage.k8s.io` API group:
+
+```golang
+type VolumeSnapshotLink struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	Spec VolumeSnapshotLinkSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+}
+
+type VolumeSnapshotLinkList struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	Items []VolumeSnapshotLink `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// VolumeSnapshotLinkSpec describes attributes of a volume snapshot link.
+type VolumeSnapshotLinkSpec struct {
+	// This field is immutable after creation.
+	Source VolumeSnapshotLinkSource `json:"source" protobuf:"bytes,1,opt,name=source"`
+}
+
+// VolumeSnapshotLinkSource specifies a reference to VolumeSnapshot.
+type VolumeSnapshotLinkSource struct {
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	Namespace string `json:"namespace" protobuf:"bytes,2,opt,name=namespace"`
+}
+```
+
+### Populator implementation
+
+The populator logic can be implemented either [(a) inside the existing CSI external-provisioner](#a-inside-the-existing-csi-external-provisioner) or [(b) as a separate populator](#b-as-a-separate-populator).
+Cluster admins can choose which implementation to be used per CSI driver basis.
+As a reference implementation, only (a) will be implemented in the community.
+
+Regardless of the implementation,
+- `VolumeSnapshotLink` CRD and `ReferencePolicy` CRD must exist in the cluster before the populator is deployed.
+- `VolumePopulator` CRD to allow popluating from `VolumeSnapshotLink` CRD needs to be created to enable this feature, as AnyVolumeDataSource feature defines. The `VolumePopulator` CRD needed for this feature will be as follows:
+```yaml
+kind: VolumePopulator
+apiVersion: populator.storage.k8s.io/v1beta1
+metadata:
+  name: volumesnapshotlink
+sourceKind:
+  group: snapshot.storage.k8s.io
+  kind: VolumeSnapshotLink
+```
+
+#### (a) inside the existing CSI external-provisioner
+
+Once populator is implemented inside the existing CSI external-provisioner, the CSI external provisioner:
+- Handles `VolumeSnapshotLink` CRD and `ReferencePolicy` CRD,
+- Checks if `VolumeSnapshotLink` is specified as `DataSourceRef`:
+  - If specified, check if the access to the `VolumeSnapshot` referenced by the `VolumeSnapshotLink` is allowed by any `ReferencePolicy`s:
+    - If allowed, use the `VolumeSnapshot` as a SnapshotSource to pass to the CSI driver for provision.
+
+To enable this feature in CSI external provisioner, `--cross-namespace-snapshot=true`
+command line flag needs to be passed to the provisioner for each CSI plugin.
+
+#### (b) as a separate populator
+
+There will be two approaches to implement as a separate populator:
```

Comment:
It's better to implement this as a separate populator in the alpha stage to avoid making external-provisioner more complicated.  When moving from alpha to beta, we can re-evaluate to see if it is better to move the logic into the external-provisioner.

Reply:
I prefer implementing by (a), because implementing (b)(2) will introduce a really complex codes as explained (it requires duplicating provisioner's provisioning codes). But, let me consider a bit more, if we can make it simple even with (b)(2).

Also, I would like to get feedback from others.

Reply:
@bswartz @xing-yang @Elbehery For further discussions, I've implemented cross-namespace snapshot populator for csi-hostpath driver in https://github.com/kubernetes-csi/lib-volume-populator/pull/31 . This is using (b)(1) approach. 
For (b)(2) approach, we may need to change CSI spec to add a new CSI call to populate the already provisioned volume, instead of only allowing creating pre-populated volume in CreateVolume (Please also see comment [here](https://github.com/kubernetes-csi/lib-volume-populator/pull/31#issuecomment-1146155307)). 

Any ideas on how we proceed for alpha?

I think that we have below choices:

1. Only define API
2. Define API and provide an example populator (for csi-hostpath) ... (b)(1) approach
3. Define API and provide common populator, which would require CSI spec change ... (b)(2) approach
4. Define API and provide a reference implementation as provisioner ... (a) approach

I still prefer 4 personally, but we may be able to choose 1 and 2 for alpha because there will be no reason to block defining API.

Reply:
Updated the KEP in implementation section (and added section on CR conflict risk). PTAL

BTW, thank you for your feedback via chat @bswartz I hope that I've added enough information.
Also, thank you for letting me know that you're under review via chat @Elbehery .